File: scripts/process_emails.py
#!/usr/bin/env python3
"""
Email processing script for Sabhya's Daily Task Tracker.
Fetches Gmail, processes with Claude AI, updates JSON task database.
"""

import os
import json
import base64
import hashlib
import re
import sys
import time
from datetime import datetime, timezone, timedelta
from email import message_from_bytes
from email.header import decode_header, make_header

import anthropic
from google.oauth2.credentials import Credentials
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError
import pytz

# ── Paths ──────────────────────────────────────────────────────────────────
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DATA_DIR = os.path.join(BASE_DIR, "data")
TASKS_FILE = os.path.join(DATA_DIR, "tasks.json")
CLIENTS_FILE = os.path.join(DATA_DIR, "clients.json")
EMPLOYEES_FILE = os.path.join(DATA_DIR, "employees.json")
SYNC_STATE_FILE = os.path.join(DATA_DIR, "sync_state.json")

# ── Constants ──────────────────────────────────────────────────────────────
GMAIL_SCOPES = ["https://www.googleapis.com/auth/gmail.readonly"]
MAX_EMAILS_PER_RUN = 200
FIRST_RUN_DAYS = 30
BATCH_SIZE = 10  # emails sent to Claude per API call


# ── Helpers ────────────────────────────────────────────────────────────────

def load_json(path: str, default: dict) -> dict:
    try:
        with open(path, "r", encoding="utf-8") as f:
            return json.load(f)
    except (FileNotFoundError, json.JSONDecodeError):
        return default


def save_json(path: str, data: dict) -> None:
    with open(path, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=2, ensure_ascii=False, default=str)


def utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def make_task_id(thread_id: str, subject: str, idx: int = 0) -> str:
    raw = f"{thread_id}:{subject}:{idx}"
    return "task_" + hashlib.md5(raw.encode()).hexdigest()[:12]


def make_client_id(name: str) -> str:
    slug = re.sub(r"[^a-z0-9]", "_", name.lower().strip())
    return "client_" + slug[:24]


def decode_mime_header(header: str) -> str:
    try:
        return str(make_header(decode_header(header or "")))
    except Exception:
        return header or ""


# ── Gmail Auth ─────────────────────────────────────────────────────────────

def get_gmail_service():
    creds = Credentials(
        token=None,
        refresh_token=os.environ["GMAIL_REFRESH_TOKEN"],
        client_id=os.environ["GMAIL_CLIENT_ID"],
        client_secret=os.environ["GMAIL_CLIENT_SECRET"],
        token_uri="https://oauth2.googleapis.com/token",
        scopes=GMAIL_SCOPES,
    )
    return build("gmail", "v1", credentials=creds, cache_discovery=False)


# ── Email Fetching ─────────────────────────────────────────────────────────

def get_email_body(payload: dict) -> str:
    """Recursively extract plain text body from Gmail message payload."""
    parts = payload.get("parts", [])
    if not parts:
        data = payload.get("body", {}).get("data", "")
        if data:
            return base64.urlsafe_b64decode(data + "==").decode("utf-8", errors="replace")
        return ""

    text = ""
    for part in parts:
        mime = part.get("mimeType", "")
        if mime == "text/plain":
            data = part.get("body", {}).get("data", "")
            if data:
                text += base64.urlsafe_b64decode(data + "==").decode("utf-8", errors="replace")
        elif mime.startswith("multipart/"):
            text += get_email_body(part)

    return text.strip()


def fetch_thread_messages(service, thread_id: str) -> list:
    """Fetch all messages in a Gmail thread."""
    try:
        thread = service.users().threads().get(userId="me", id=thread_id, format="full").execute()
        messages = []
        for msg in thread.get("messages", []):
            headers = {h["name"]: h["value"] for h in msg.get("payload", {}).get("headers", [])}
            body = get_email_body(msg.get("payload", {}))
            messages.append({
                "id": msg["id"],
                "date": headers.get("Date", ""),
                "from": decode_mime_header(headers.get("From", "")),
                "to": decode_mime_header(headers.get("To", "")),
                "subject": decode_mime_header(headers.get("Subject", "")),
                "body": body[:3000],  # cap to avoid token overflow
            })
        return messages
    except HttpError as e:
        print(f"  Warning: could not fetch thread {thread_id}: {e}")
        return []


def list_new_threads(service, sync_state: dict, force_full: bool) -> list:
    """Return list of thread IDs to process, respecting sync state."""
    query_parts = ["in:inbox OR in:sent"]

    if force_full or not sync_state.get("first_run_completed"):
        cutoff = datetime.now(timezone.utc) - timedelta(days=FIRST_RUN_DAYS)
        query_parts.append(f"after:{int(cutoff.timestamp())}")
        print(f"First-run / full-sync mode: fetching emails from last {FIRST_RUN_DAYS} days")
    elif sync_state.get("last_sync_time"):
        cutoff = datetime.fromisoformat(sync_state["last_sync_time"]) - timedelta(minutes=5)
        query_parts.append(f"after:{int(cutoff.timestamp())}")
        print(f"Incremental sync since: {cutoff.isoformat()}")

    query = " ".join(query_parts)
    thread_ids = []
    page_token = None

    while True:
        try:
            params = {"userId": "me", "q": query, "maxResults": 100}
            if page_token:
                params["pageToken"] = page_token
            resp = service.users().threads().list(**params).execute()
            threads = resp.get("threads", [])
            thread_ids.extend(t["id"] for t in threads)
            page_token = resp.get("nextPageToken")
            if not page_token or len(thread_ids) >= MAX_EMAILS_PER_RUN:
                break
        except HttpError as e:
            print(f"Error listing threads: {e}")
            break

    return thread_ids[:MAX_EMAILS_PER_RUN]


# ── Claude Processing ──────────────────────────────────────────────────────

SYSTEM_PROMPT = """You are an intelligent email-to-task extraction assistant for a professional task tracker.
Your job is to analyze email threads and extract structured task information.

Rules:
1. Identify the CLIENT NAME from email domain, contact name, company references, or subject matter.
   - If no clear client, use "Internal" as client name.
2. Extract all ACTIONABLE ITEMS as individual tasks.
3. Determine PRIORITY: urgent (deadlines within 48h or explicit urgent language), medium (general work items), low (informational/FYI).
4. Determine COMPLETION STATUS: if latest emails indicate the matter is resolved/done/closed, mark as completed.
5. Identify the PERSON RESPONSIBLE for next steps (either the user or someone mentioned in the email).
6. Write a SUMMARY of the full thread conversation in 2-4 sentences.
7. For TASK TITLE: make it concise and action-oriented (max 80 chars).
8. Return ONLY valid JSON. No explanation text outside JSON.

Output format:
{
  "client_name": "string",
  "tasks": [
    {
      "title": "string",
      "priority": "urgent|medium|low",
      "status": "pending|completed",
      "responsible_person": "string",
      "actionables": ["string", ...],
      "thread_summary": "string",
      "due_date_hint": "string or null"
    }
  ]
}
"""


def process_thread_with_claude(client: anthropic.Anthropic, thread_messages: list) -> dict | None:
    """Send thread to Claude and parse the structured task response."""
    if not thread_messages:
        return None

    subject = thread_messages[0].get("subject", "(no subject)")
    thread_text = f"Email Thread: {subject}\n\n"
    for i, msg in enumerate(thread_messages, 1):
        thread_text += (
            f"--- Message {i} ---\n"
            f"From: {msg['from']}\n"
            f"To: {msg['to']}\n"
            f"Date: {msg['date']}\n"
            f"Body:\n{msg['body']}\n\n"
        )

    try:
        response = client.messages.create(
            model="claude-sonnet-4-6",
            max_tokens=2048,
            system=SYSTEM_PROMPT,
            messages=[{"role": "user", "content": thread_text}],
        )
        raw = response.content[0].text.strip()

        # Strip markdown code fences if present
        raw = re.sub(r"^```(?:json)?\s*", "", raw)
        raw = re.sub(r"\s*```$", "", raw)

        return json.loads(raw)
    except (json.JSONDecodeError, anthropic.APIError, IndexError) as e:
        print(f"  Claude error: {e}")
        return None


# ── Database Updates ───────────────────────────────────────────────────────

def upsert_client(clients_db: dict, client_name: str) -> str:
    """Return client_id, creating client if it doesn't exist."""
    client_id = make_client_id(client_name)
    existing = {c["id"]: c for c in clients_db.get("clients", [])}

    if client_id not in existing:
        clients_db.setdefault("clients", []).append({
            "id": client_id,
            "name": client_name,
            "created_at": utc_now_iso(),
            "order": len(clients_db["clients"]),
        })
    return client_id


def upsert_task(tasks_db: dict, task_data: dict, thread_id: str,
                thread_messages: list, client_id: str, client_name: str) -> None:
    """Insert or update a task in the database."""
    subject = thread_messages[0].get("subject", "") if thread_messages else ""
    task_id = make_task_id(thread_id, task_data["title"])

    existing_map = {t["id"]: t for t in tasks_db.get("tasks", [])}

    email_refs = [m["id"] for m in thread_messages]
    thread_summary_msgs = [
        {"from": m["from"], "date": m["date"], "snippet": m["body"][:500]}
        for m in thread_messages
    ]

    if task_id in existing_map:
        task = existing_map[task_id]
        # Update fields — never overwrite manual user changes to priority/assignee
        task["thread_summary"] = task_data.get("thread_summary", task.get("thread_summary", ""))
        task["actionables"] = task_data.get("actionables", task.get("actionables", []))
        task["responsible_person"] = task_data.get("responsible_person", task.get("responsible_person", ""))
        task["email_refs"] = list(set(task.get("email_refs", []) + email_refs))
        task["thread_messages"] = thread_summary_msgs
        task["updated_at"] = utc_now_iso()
        # Auto-complete only if Claude is confident
        if task_data.get("status") == "completed" and task["status"] == "pending":
            task["status"] = "completed"
            task["completed_at"] = utc_now_iso()
    else:
        new_task = {
            "id": task_id,
            "title": task_data["title"],
            "priority": task_data.get("priority", "medium"),
            "status": task_data.get("status", "pending"),
            "client_id": client_id,
            "client_name": client_name,
            "responsible_person": task_data.get("responsible_person", ""),
            "assigned_to": "",
            "thread_summary": task_data.get("thread_summary", ""),
            "actionables": task_data.get("actionables", []),
            "due_date_hint": task_data.get("due_date_hint"),
            "email_refs": email_refs,
            "thread_messages": thread_summary_msgs,
            "thread_subject": subject,
            "thread_id": thread_id,
            "created_at": utc_now_iso(),
            "updated_at": utc_now_iso(),
            "completed_at": None,
            "manually_edited": False,
        }
        tasks_db.setdefault("tasks", []).append(new_task)


# ── Main Orchestration ─────────────────────────────────────────────────────

def main():
    print(f"[{utc_now_iso()}] Starting email sync...")

    # Check required secrets
    required_env = ["GMAIL_CLIENT_ID", "GMAIL_CLIENT_SECRET", "GMAIL_REFRESH_TOKEN", "ANTHROPIC_API_KEY"]
    missing = [e for e in required_env if not os.environ.get(e)]
    if missing:
        print(f"ERROR: Missing environment variables: {', '.join(missing)}")
        print("Please add these as GitHub repository secrets.")
        sys.exit(1)

    force_full = os.environ.get("FORCE_FULL_SYNC", "false").lower() == "true"

    # Load databases
    tasks_db = load_json(TASKS_FILE, {"version": "1.0", "last_updated": "", "tasks": []})
    clients_db = load_json(CLIENTS_FILE, {"version": "1.0", "clients": []})
    sync_state = load_json(SYNC_STATE_FILE, {
        "version": "1.0", "last_sync_time": "", "last_message_id": "",
        "total_emails_processed": 0, "first_run_completed": False, "sync_history": []
    })

    # Connect to Gmail
    print("Connecting to Gmail...")
    service = get_gmail_service()

    # Connect to Claude
    claude = anthropic.Anthropic(api_key=os.environ["ANTHROPIC_API_KEY"])

    # Get thread IDs to process
    thread_ids = list_new_threads(service, sync_state, force_full)
    print(f"Found {len(thread_ids)} threads to process")

    if not thread_ids:
        print("No new threads. Updating sync timestamp.")
        sync_state["last_sync_time"] = utc_now_iso()
        save_json(SYNC_STATE_FILE, sync_state)
        return

    processed = 0
    errors = 0

    for i, thread_id in enumerate(thread_ids, 1):
        print(f"  [{i}/{len(thread_ids)}] Processing thread {thread_id}...")
        try:
            messages = fetch_thread_messages(service, thread_id)
            if not messages:
                continue

            result = process_thread_with_claude(claude, messages)
            if not result:
                errors += 1
                continue

            client_name = result.get("client_name", "Unknown")
            client_id = upsert_client(clients_db, client_name)

            for task_data in result.get("tasks", []):
                upsert_task(tasks_db, task_data, thread_id, messages, client_id, client_name)

            processed += 1

            # Respect rate limits
            if i % 10 == 0:
                time.sleep(1)

        except Exception as e:
            print(f"  Error processing thread {thread_id}: {e}")
            errors += 1
            continue

    # Update metadata
    now = utc_now_iso()
    tasks_db["last_updated"] = now
    sync_state["last_sync_time"] = now
    sync_state["total_emails_processed"] = sync_state.get("total_emails_processed", 0) + processed
    sync_state["first_run_completed"] = True
    sync_state.setdefault("sync_history", []).append({
        "time": now,
        "threads_found": len(thread_ids),
        "threads_processed": processed,
        "errors": errors,
    })
    # Keep only last 100 history entries
    sync_state["sync_history"] = sync_state["sync_history"][-100:]

    # Save all databases
    save_json(TASKS_FILE, tasks_db)
    save_json(CLIENTS_FILE, clients_db)
    save_json(SYNC_STATE_FILE, sync_state)

    print(f"\nSync complete: {processed} threads processed, {errors} errors")
    print(f"Total tasks in DB: {len(tasks_db.get('tasks', []))}")
    print(f"Total clients: {len(clients_db.get('clients', []))}")


if __name__ == "__main__":
    main()

File: scripts/process-emails.js
#!/usr/bin/env node
/* ══════════════════════════════════════════════════════════════════
   process-emails.js
   Runs as a scheduled GitHub Actions job every 10 minutes.
   1. Fetches new Gmail emails since last run
   2. Analyses each thread with Claude AI
   3. Creates / updates / completes tasks in data/tasks.json
   4. Commits the updated file back to the repo

   Required environment variables (set as GitHub Secrets):
     GMAIL_CLIENT_ID       — Google OAuth2 client ID
     GMAIL_CLIENT_SECRET   — Google OAuth2 client secret
     GMAIL_REFRESH_TOKEN   — Gmail OAuth2 refresh token
     ANTHROPIC_API_KEY     — Claude API key
     GITHUB_TOKEN          — Auto-provided by GitHub Actions
     GITHUB_REPOSITORY     — Auto-provided (owner/repo)
   ══════════════════════════════════════════════════════════════════ */

'use strict';

const fs        = require('fs');
const path      = require('path');
const { google }   = require('googleapis');
const Anthropic    = require('@anthropic-ai/sdk');
const { Octokit }  = require('@octokit/rest');

/* ── Configuration ── */
const cfg = {
  gmail: {
    clientId:     process.env.GMAIL_CLIENT_ID,
    clientSecret: process.env.GMAIL_CLIENT_SECRET,
    refreshToken: process.env.GMAIL_REFRESH_TOKEN,
  },
  anthropic: {
    apiKey: process.env.ANTHROPIC_API_KEY,
  },
  github: {
    token:    process.env.GITHUB_TOKEN,
    repo:     process.env.GITHUB_REPOSITORY || '',        // "owner/repo"
    branch:   process.env.GITHUB_REF_NAME  || 'main',
    filePath: 'data/tasks.json',
  }
};

const DATA_FILE = path.join(__dirname, '..', 'data', 'tasks.json');

/* ── Helpers ── */

function uuid() {
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
    const r = Math.random() * 16 | 0;
    return (c === 'x' ? r : (r & 0x3 | 0x8)).toString(16);
  });
}

function log(msg) { console.log(`[${new Date().toISOString()}] ${msg}`); }

function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

/* ── Load / save task data ── */

function loadData() {
  if (!fs.existsSync(DATA_FILE)) {
    return {
      metadata: { version: '1.0.0', lastUpdated: null, lastEmailFetched: null,
                  totalEmailsProcessed: 0, initialized: false },
      clients: {}, tasks: [], employees: [], syncLog: []
    };
  }
  return JSON.parse(fs.readFileSync(DATA_FILE, 'utf8'));
}

function saveData(data) {
  data.metadata.lastUpdated = new Date().toISOString();
  fs.writeFileSync(DATA_FILE, JSON.stringify(data, null, 2));
}

/* ── Gmail ── */

function buildGmailClient() {
  const auth = new google.auth.OAuth2(cfg.gmail.clientId, cfg.gmail.clientSecret);
  auth.setCredentials({ refresh_token: cfg.gmail.refreshToken });
  return google.gmail({ version: 'v1', auth });
}

async function fetchEmailsSince(gmail, sinceDate) {
  const q = sinceDate
    ? `after:${Math.floor(new Date(sinceDate).getTime() / 1000)}`
    : `newer_than:30d`;

  log(`Fetching emails with query: ${q}`);

  const threads = [];
  let pageToken;

  do {
    const res = await gmail.users.threads.list({
      userId: 'me', q, maxResults: 50, pageToken
    });

    const items = res.data.threads || [];
    for (const item of items) {
      const thread = await gmail.users.threads.get({
        userId: 'me', id: item.id, format: 'full'
      });
      threads.push(thread.data);
      await sleep(50); // Avoid rate limits
    }

    pageToken = res.data.nextPageToken;
  } while (pageToken && threads.length < 200);

  log(`Fetched ${threads.length} threads`);
  return threads;
}

function extractBody(payload) {
  if (!payload) return '';

  function decode(data) {
    try { return Buffer.from(data, 'base64url').toString('utf8'); }
    catch { return ''; }
  }

  if (payload.body?.data) return decode(payload.body.data);

  if (payload.parts) {
    for (const part of payload.parts) {
      if (part.mimeType === 'text/plain' && part.body?.data)
        return decode(part.body.data);
    }
    for (const part of payload.parts) {
      const nested = extractBody(part);
      if (nested) return nested;
    }
  }
  return '';
}

function parseMessage(msg) {
  const headers = {};
  (msg.payload?.headers || []).forEach(h => { headers[h.name.toLowerCase()] = h.value; });

  return {
    id:      msg.id,
    from:    headers['from']    || '',
    to:      headers['to']      || '',
    subject: headers['subject'] || '',
    date:    headers['date']    || new Date(parseInt(msg.internalDate)).toISOString(),
    snippet: msg.snippet || '',
    body:    extractBody(msg.payload).substring(0, 2000), // truncate for API
  };
}

function parseThread(thread) {
  const messages = (thread.messages || []).map(parseMessage);
  return {
    id:       thread.id,
    messages,
    subject:  messages[0]?.subject || '',
    latest:   messages[messages.length - 1],
    earliest: messages[0],
  };
}

/* ── Claude AI ── */

function buildAnalysisClient() {
  return new Anthropic({ apiKey: cfg.anthropic.apiKey });
}

async function analyseThread(client, thread, existingTask) {
  const emailSummary = thread.messages.map((m, i) =>
    `Email ${i + 1} [${m.date}] From: ${m.from}\nSubject: ${m.subject}\n${m.body || m.snippet}`
  ).join('\n\n---\n\n');

  const systemPrompt = `You are an intelligent task management assistant. You analyse email threads and extract actionable task information. Always respond with valid JSON only. Today's date: ${new Date().toISOString().split('T')[0]}.`;

  const userPrompt = `Analyse this email thread and extract task information. Return a JSON object with exactly these fields:

{
  "clientName": "the company/person this email relates to (not the sender's company, but the client/project)",
  "title": "concise task title (max 80 chars)",
  "summary": "2-3 sentence summary of the entire thread and current situation",
  "priority": "urgent|medium|low — urgent if deadline within 3 days or marked urgent/ASAP, low if informational only",
  "actionables": ["list", "of", "specific", "action", "items", "needed"],
  "nextStepsResponsible": "name or role of person responsible for next action",
  "isCompleted": true or false — true if the thread shows the matter is resolved/closed/completed,
  "confidence": 0.0-1.0 — how confident you are this is a real actionable task (not spam/newsletter)
}

${existingTask ? `\nExisting task for context:\nTitle: ${existingTask.title}\nClient: ${existingTask.clientName}\nPriority: ${existingTask.priority}\n` : ''}

Email Thread:
${emailSummary}

Respond with ONLY the JSON object, no markdown, no explanation.`;

  let attempts = 0;
  while (attempts < 3) {
    try {
      const response = await client.messages.create({
        model: 'claude-sonnet-4-6',
        max_tokens: 1024,
        messages: [{ role: 'user', content: userPrompt }],
        system: systemPrompt,
      });

      const text = response.content[0]?.text?.trim() || '{}';
      const json = text.startsWith('{') ? text : text.substring(text.indexOf('{'), text.lastIndexOf('}') + 1);
      return JSON.parse(json);
    } catch (err) {
      attempts++;
      log(`Claude error (attempt ${attempts}): ${err.message}`);
      if (attempts < 3) await sleep(2000 * attempts);
    }
  }

  return null;
}

/* ── Task matching ── */

function findMatchingTask(tasks, thread, analysis) {
  // Match by email thread ID reference
  for (const t of tasks) {
    if ((t.emailThread || []).some(e => e.id === thread.id)) return t;
  }

  // Match by Gmail thread ID stored in task
  if (tasks.some(t => t.gmailThreadId === thread.id)) {
    return tasks.find(t => t.gmailThreadId === thread.id);
  }

  // Fuzzy match by client + similar title
  const clientName = (analysis?.clientName || '').toLowerCase();
  const titleWords = (analysis?.title || '').toLowerCase().split(' ').filter(w => w.length > 4);

  return tasks.find(t => {
    if (t.status === 'completed') return false;
    const sameClient = t.clientName?.toLowerCase() === clientName;
    if (!sameClient) return false;
    const taskWords = (t.title || '').toLowerCase().split(' ');
    const matches = titleWords.filter(w => taskWords.some(tw => tw.includes(w) || w.includes(tw)));
    return matches.length >= Math.ceil(titleWords.length * 0.5);
  });
}

function ensureClient(data, clientName) {
  const existing = Object.values(data.clients).find(
    c => c.name.toLowerCase() === clientName.toLowerCase()
  );
  if (existing) return existing;

  const client = { id: uuid(), name: clientName, createdAt: new Date().toISOString() };
  data.clients[client.id] = client;
  return client;
}

/* ── Process a single thread ── */

async function processThread(aiClient, thread, data) {
  log(`Processing thread: ${thread.subject} (${thread.messages.length} messages)`);

  const existing = findMatchingTask(data.tasks, thread, null);
  const analysis = await analyseThread(aiClient, thread, existing);

  if (!analysis) { log('  → Skipped (AI error)'); return; }
  if (analysis.confidence < 0.4) { log(`  → Skipped (low confidence: ${analysis.confidence})`); return; }

  const clientName  = analysis.clientName || 'General';
  const clientRecord = ensureClient(data, clientName);

  const threadEmails = thread.messages.map(m => ({
    id:      m.id,
    from:    m.from,
    to:      m.to,
    subject: m.subject,
    date:    m.date,
    snippet: m.snippet,
    body:    m.body,
  }));

  if (existing) {
    // Update existing task
    log(`  → Updating task: "${existing.title}"`);

    // Don't override user-set priority or assignee
    if (!existing._userPriority) existing.priority = analysis.priority;
    existing.summary              = analysis.summary;
    existing.actionables          = analysis.actionables || [];
    existing.nextStepsResponsible = analysis.nextStepsResponsible;
    existing.updatedAt            = new Date().toISOString();
    existing.gmailThreadId        = thread.id;

    // Merge email thread (add new messages, don't duplicate)
    const existingIds = new Set((existing.emailThread || []).map(e => e.id));
    for (const email of threadEmails) {
      if (!existingIds.has(email.id)) existing.emailThread.push(email);
    }

    // Auto-complete if AI says it's done
    if (analysis.isCompleted && !existing._userStatus) {
      existing.status      = 'completed';
      existing.completedAt = new Date().toISOString();
      log(`  → Auto-completed task`);
    }
  } else {
    // Create new task
    log(`  → Creating task: "${analysis.title}" [${analysis.priority}] for ${clientName}`);

    const task = {
      id:                   uuid(),
      title:                analysis.title,
      clientId:             clientRecord.id,
      clientName:           clientRecord.name,
      priority:             analysis.priority || 'medium',
      status:               analysis.isCompleted ? 'completed' : 'pending',
      source:               'email',
      gmailThreadId:        thread.id,
      summary:              analysis.summary,
      description:          analysis.summary,
      actionables:          analysis.actionables || [],
      nextStepsResponsible: analysis.nextStepsResponsible,
      assignedTo:           null,
      emailThread:          threadEmails,
      createdAt:            new Date().toISOString(),
      updatedAt:            new Date().toISOString(),
      completedAt:          analysis.isCompleted ? new Date().toISOString() : null,
    };

    data.tasks.push(task);
  }
}

/* ── GitHub commit ── */

async function commitToGitHub(data) {
  const [owner, repo] = cfg.github.repo.split('/');
  if (!owner || !repo) { log('No GitHub repo configured — skipping commit'); return; }

  const octokit = new Octokit({ auth: cfg.github.token });
  const content = Buffer.from(JSON.stringify(data, null, 2)).toString('base64');

  let sha;
  try {
    const { data: fileData } = await octokit.repos.getContent({
      owner, repo, path: cfg.github.filePath, ref: cfg.github.branch
    });
    sha = fileData.sha;
  } catch (e) {
    if (e.status !== 404) throw e;
  }

  await octokit.repos.createOrUpdateFileContents({
    owner, repo, branch: cfg.github.branch,
    path:    cfg.github.filePath,
    message: `sync: email processor update [${new Date().toISOString()}]\n\nProcessed ${data.metadata.totalEmailsProcessed} emails total`,
    content,
    ...(sha ? { sha } : {})
  });

  log('Committed tasks.json to GitHub');
}

/* ── Main ── */

async function main() {
  log('=== Email Processor Starting ===');

  // Validate required env vars
  const required = ['GMAIL_CLIENT_ID', 'GMAIL_CLIENT_SECRET', 'GMAIL_REFRESH_TOKEN', 'ANTHROPIC_API_KEY'];
  const missing  = required.filter(k => !process.env[k]);
  if (missing.length > 0) {
    log(`ERROR: Missing environment variables: ${missing.join(', ')}`);
    log('Please configure these as GitHub Secrets. See README for setup instructions.');
    process.exit(1);
  }

  const data     = loadData();
  const gmail    = buildGmailClient();
  const aiClient = buildAnalysisClient();

  const isFirstRun = !data.metadata.initialized;
  const since      = isFirstRun ? null : data.metadata.lastEmailFetched;

  if (isFirstRun) log('First run: fetching last 30 days of emails');

  // Fetch email threads
  let threads;
  try {
    threads = await fetchEmailsSince(gmail, since);
  } catch (err) {
    log(`Gmail fetch error: ${err.message}`);
    process.exit(1);
  }

  if (threads.length === 0) {
    log('No new emails. Nothing to process.');
    data.metadata.lastEmailFetched = new Date().toISOString();
    saveData(data);
    await commitToGitHub(data);
    return;
  }

  // Process each thread (with concurrency limit of 3)
  const parsed = threads.map(parseThread);
  const CONCURRENCY = 3;

  for (let i = 0; i < parsed.length; i += CONCURRENCY) {
    const batch = parsed.slice(i, i + CONCURRENCY);
    await Promise.all(batch.map(t => processThread(aiClient, t, data)));
    await sleep(1000); // Respect Claude rate limits
  }

  data.metadata.lastEmailFetched      = new Date().toISOString();
  data.metadata.totalEmailsProcessed += threads.length;
  data.metadata.initialized           = true;

  // Record sync log entry
  data.syncLog = (data.syncLog || []).slice(-100); // keep last 100 entries
  data.syncLog.push({
    timestamp:  new Date().toISOString(),
    threadsProcessed: threads.length,
    tasksTotal: data.tasks.length,
    isFirstRun,
  });

  saveData(data);
  log(`Saved data: ${data.tasks.length} total tasks`);

  await commitToGitHub(data);

  log('=== Done ===');
}

main().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});
